onnxruntime/core/providers/cpu/nn/tfidfvectorizer.cc

```diff
@@ -490,104 +414,15 @@ Status TfIdfVectorizer::ComputeImpl(OpKernelContext* ctx) const {
     return Status::OK();
   }
 
-  assert((b_dim * C) == total_items);
+  std::function<void(int32_t)> fn = [this, ctx, C, &frequencies](int32_t row_num) {
+    ComputeImpl(ctx, row_num, C, frequencies);
+  };
 
-  const auto max_gram_length = impl.max_gram_length_;
-  const auto max_skip_distance = impl.max_skip_count_ + 1;  // Convert to distance
-  auto start_ngram_size = impl.min_gram_length_;
-  auto const input_data = X->template Data<T>();
-  auto const end_data = input_data + total_items;
-  NgramEntry<T> sample;
-
-  // Treat 1-grams in a special way
-  if (start_ngram_size == 1) {
-    size_t row_num = 0;
-    auto ngram_start = input_data;
-    while (ngram_start < end_data) {
-      auto const ngram_row_end = ngram_start + C;
-      while (ngram_start < ngram_row_end) {
-        sample.Clear();
-        sample.AddItem(*ngram_start);
-        auto hit = impl.PoolFind<T>(sample);
-        if (hit != set_end) {
-          // record frequency
-          auto ngram_id = hit->Id();
-          impl.IncrementCount(ngram_id, row_num, frequencies);
-        }
-        ++ngram_start;
-      }
-      ++row_num;
-      ngram_start = ngram_row_end;
-    }
-    if (++start_ngram_size > max_gram_length) {
-      OutputResult(ctx, B, frequencies);
-      return Status::OK();
-    }
-  }
+  concurrency::ThreadPool::TryBatchParallelFor(ctx->GetOperatorThreadPool(), num_rows, std::move(fn), num_rows / 3);
 
```

Comment:
3 rows per batch seems small. How much work is involved in processing a row? 

It would be good to determine the number of batches based on a) number of threads available and b) amount of work per batch. Not sure there's any point having more batches than threads. Also not good to use lots of threads if each only has a small bit of work to do.

Reply:
It depends on a few factors. The sizes of n-grams the user interested and the size of the row. I would say it would be ngram_sizes * row_size. With the recent change it is hopefully closer to row_size.

---
In reply to: [391959114](https://github.com/microsoft/onnxruntime/pull/3197#discussion_r391959114) [](ancestors = 391959114)

Reply:
Re-calculated the number of batches so it is never more than the number of threads.

---
In reply to: [391970262](https://github.com/microsoft/onnxruntime/pull/3197#discussion_r391970262) [](ancestors = 391970262,391959114)

Reply:
TrybatchParallelFor() will calculate the correct number of batches.

---
In reply to: [391989799](https://github.com/microsoft/onnxruntime/pull/3197#discussion_r391989799) [](ancestors = 391989799,391970262,391959114)

Reply:
"not good to use lots of threads if each only has a small bit of work to do."
Indeed this part is not solved yet. Let's leave it as a TODO.